### PR TITLE
Fix user and group determination

### LINKED ISSUE #18: Users sometimes not found
Two occurrences:

` File \"sagemaker_shim/models.py\", line 302, in _get_user_info`

RuntimeError: User 1000 not found

RuntimeError: User 1001 not found

First one is from @pkcakeout, have requested the Dockerfile.

### PR DESCRIPTION
**Pull request by jmsmkn**
Closes #18

### CODE REVIEW

**Review by pkcakeout — COMMENTED** [review]
Looks good, I think!

Could not find the Popen line to suggest the modification from https://github.com/DIAGNijmegen/rse-sagemaker-shim/issues/18#issuecomment-1884856973 though. Maybe that should be added, as you suggested!

**Comment by jmsmkn** [comment]
I explicitly set `extra_groups=[]` in `0b37f5d`. Should that be set to anything else in other situations? We do a lookup of the user in `pwd` so adding another parameter is no problem.

**Comment by pkcakeout** [comment]
> I explicitly set extra_groups=[] in 0b37f5d. Should that be set to anything else in other situations? We do a lookup of the user in pwd so adding another parameter is no problem.

Yeah, I do not know. If this was a full Linux system, there will be issues from not assigning supplemental groups. However, within docker, I do not expect issues to appear...

**Comment by pkcakeout** [comment]
image

Is there a race condition in the testing framework?!

**Comment by pkcakeout** [comment]
Anyway... there seem to be issues with dropping the extra groups in sagemaker afterall... in the commandline you can run:

~~~~~~~~~~~
paul@dingo-linux:~$ groups paul
paul : paul adm cdrom sudo dip plugdev lpadmin lxd sambashare docker libvirt
~~~~~~~~~~~

that list (not this literal one, of course!) should be the list of supplemental groups for the user. It is only available when the user actually exists.

**Comment by pkcakeout** [comment]
... `id` is actually still the simpler option:

~~~~~~~
paul@dingo-linux:~$ id 1000
uid=1000(paul) gid=1000(paul) groups=1000(paul),4(adm),24(cdrom),27(sudo),30(dip),46(plugdev),122(lpadmin),132(lxd),133(sambashare),999(docker),141(libvirt)
~~~~~~~

Forgot that it also nicely lists the "groups=..." bit.

File: sagemaker_shim/models.py [hunk 1]
```diff
@@ -288,61 +354,27 @@ def proc_env(self) -> dict[str, str]:
...
-            else:
-                gid = group.gr_gid
-
-            return UserGroup(uid=uid, gid=gid, home=home)
+            info = _get_user_info(id_or_name=match.group("user"))
+            gid = _get_group_id(id_or_name=match.group("group"))
+
+            return UserInfo(
+                uid=info.uid,
+                gid=info.gid if gid is None else gid,
```
Comment (jmsmkn) [inline sagemaker_shim/models.py:372]: Here we set an explicit `gid` for the process (`group=gid` in Popen). Does that `gid` also need to be added to `groups` if it is not there already?

Comment (pkcakeout) [reply sagemaker_shim/models.py:372]: No, not that I know. The supplemental groups are handled somehow differently from the uid/gid. Though I do not quite know how those work either. I quickly tried to find out now by inspecting the `/proc/[prog-id]` directory, but I do not think that the supplemental groups are listed there. The gid is... no clue. The supplemental/extra groups must be resolved using some other mechanism...

Comment (pkcakeout) [reply sagemaker_shim/models.py:372]: Well, digging further makes me question my earlier answer:

https://www.man7.org/linux/man-pages/man2/getgroups.2.html

> It is unspecified whether the effective group ID of the calling process is included in the returned list.  (Thus, an application should also call [getegid(2)](https://www.man7.org/linux/man-pages/man2/getegid.2.html) and add or remove the resulting value.)

Great! So maybe we should add it just-in-case?

Comment (jmsmkn) [reply sagemaker_shim/models.py]: Done!

File: tests/test_models.py [hunk 2]
```diff
@@ -97,20 +103,26 @@ def test_removing_ld_library_path(monkeypatch):
...
         ),
         # User does not exist, but is an int
         ("23746", 23746, None, None, []),
-        (f"23746:{grp.getgrgid(0).gr_name}", 23746, 0, None, []),
-        (f"23746:{os.getgid()}", 23746, os.getgid(), None, []),
+        (f"23746:{grp.getgrgid(0).gr_name}", 23746, 0, None, [0]),
+        (f"23746:{os.getgid()}", 23746, os.getgid(), None, [os.getgid()]),
         # User and group do not exist, but are ints
-        ("23746:23746", 23746, 23746, None, []),
+        ("23746:23746", 23746, 23746, None, [23746]),
```
Comment (pkcakeout) [inline tests/test_models.py]: Neat!

**Pull request merged**